src/main/scala/tech/cryptonomic/conseil/tezos/TezosPlatformDiscoveryOperations.scala

```diff
@@ -391,28 +391,27 @@ class TezosPlatformDiscoveryOperations(
 
   /** Starts initialization of attributes count cache */
   def initAttributesCache: Future[Unit] = {
-    val result = for {
+    for {
       _ <- caching.updateCachingStatus(InProgress)
       entCache <- caching.getEntities(networkName)
       attributes <- caching.getAllAttributes
       _ <- contextShift.shift
-      updatedAttributes <- entCache.fold(IO(Map.empty[String, List[Attribute]]))(
-        entC => getAllUpdatedAttributes(entC.value, attributes)
+      updatedAttributes <- entCache.foldMapM(
```

Comment:
Could you explain briefly the change you made here?

Reply:
The original code used fold on an Option to define what to do 
- when the value was there: get the attributes, returning a map wrapped in IO
- when the value was missing: return an empty map wrapped in IO

`foldMap` would do the exact same, using the fact that a Map has a Monoid instance that defines an empty value, so you won't need to explicitly say what it is, it's the most natural one.

`foldMapM` takes into account that the mapping part produced a wrapping Monad (IO), therefore it automatically matches the returned empty monoid result by wrapping it into the needed monad.

It's essentially doing grunt work for free, adding what is expected to be the most natural empty value, instead of needing from the outside.

Reply:
To complete the picture, the original code then mapped on such IO to cache each entry of the map into the cache (using putAttributes(...)).

Each of those produced a Unit value. Hence the original `result` was a `Seq[IO[Unit]]` wrapped in a extra IO, which was useless, unless you sequenced those into a single IO value to be run.
Thus the `flatMap(_.toList.sequence)` action.

I simply used `traverse` to do the same within the same for comprehension, ignoring the useless result, and finally adding the last IO operation (originally `>> caching.updateCachingStatus`) to the same comprehension again.

I also removed the conversion to a `Map` and back `.toList` by using a `List` of key/value directly, since the Map structure wasn't actually ever used to index the entries. The list was only ever traversed linearly.